Log reindexing failures

ankikuma opened the pull request:
Log number of reindex tasks in flight during a shutdown. 

This PR aims to address ES-8850. 

Also fixing a small labeling mistake for the reindex metrics.

ankikuma committed:
Log reindexing failures

ankikuma commented:
A couple of things to consider while reviewing:
1. Right now I have set the timeout for the wait for reindex tasks to 1 min. Do we want a setting for this so we can tweak it ?
2. Right now I am exposing reindexing failures by logging a message. Do we want to use APM metrics instead ?

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

elasticsearchmachine commented:
Hi @ankikuma, I've created a changelog YAML for you.

Tim-Brooks commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -702,6 +704,52 @@ private void awaitSearchTasksComplete(TimeValue asyncSearchTimeout) {
         }
     }
 
+    private void awaitReindexTasksComplete(TimeValue asyncReindexTimeout) {
```
Can we combine some code here? Something like:

```java
    private void awaitSearchTasksComplete(TimeValue asyncSearchTimeout) {
        waitForTasks(asyncSearchTimeout, TransportSearchAction.NAME);
    }

    private void awaitReindexTasksComplete(TimeValue reindexTimeout) {
        waitForTasks(reindexTimeout, ReindexAction.NAME);
    }

    private void waitForTasks(TimeValue timeout, String taskName) {
        TaskManager taskManager = injector.getInstance(TransportService.class).getTaskManager();
        long millisWaited = 0;
        while (true) {
            long searchTasksRemaining = taskManager.getTasks().values().stream().filter(task -> taskName.equals(task.getAction())).count();
            if (searchTasksRemaining == 0) {
                logger.debug("all search tasks complete");
                return;
            } else {
                // Let the system work on those searches for a while. We're on a dedicated thread to manage app shutdown, so we
                // literally just want to wait and not take up resources on this thread for now. Poll period chosen to allow short
                // response times, but checking the tasks list is relatively expensive, and we don't want to waste CPU time we could
                // be spending on finishing those searches.
                final TimeValue pollPeriod = TimeValue.timeValueMillis(500);
                millisWaited += pollPeriod.millis();
                if (TimeValue.ZERO.equals(timeout) == false && millisWaited >= timeout.millis()) {
                    logger.warn(
                        format(
                            "timed out after waiting [%s] for [%d] " + taskName + " tasks to finish",
                            timeout.toString(),
                            searchTasksRemaining
                        )
                    );
                    return;
                }
                logger.debug(
                    format("waiting for [%s] " + taskName + " tasks to finish, next poll in [%s]", searchTasksRemaining, pollPeriod)
                );
                try {
                    Thread.sleep(pollPeriod.millis());
                } catch (InterruptedException ex) {
                    logger.warn(
                        format(
                            "interrupted while waiting [%s] for [%d] " + taskName + " tasks to finish",
                            timeout.toString(),
                            searchTasksRemaining
                        )
                    );
                    return;
                }
            }
        }
    }
```

Tim-Brooks reviewed (commented):
For a test I would create an `IT` performing reindex on a specific node. And then wait for the task to appear in the tasks list. And then immediately stop the node. And check that the task still completes.

You would want to try it out without your shutdown wait code here and make sure you are indexing enough data that the test would (at least sometimes) fail without your shutdown wait you are adding here.

Tim-Brooks commented:
If it is difficult to get the reindex task to fail from restarting the node, you can always reach in and grab `ThreadPool` and block some of the threads. But I would start by trying to do the test without that.

ankikuma commented:
Thanks Tim, I will work on a test case.

henningandersen commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -629,6 +630,7 @@ boolean isIncomplete() {
 
             stopperRunner.accept("http-server-transport-stop", injector.getInstance(HttpServerTransport.class)::close);
             stopperRunner.accept("async-search-stop", () -> awaitSearchTasksComplete(maxTimeout));
+            stopperRunner.accept("reindex-stop", () -> awaitReindexTasksComplete(TimeValue.min(TimeValue.timeValueMinutes(1), maxTimeout)));
```
Let us make the timeout configurable separately from the max-timeout. I think 1 min could be too long by default, I'd go with something like 10secs and we can increase where we need it. The problem is that users do rolling restarts and this could prevent those from completing as fast as before - and the longer it takes to restart the more operations may need to be replayed to re-establish sync'ed replicas.

henningandersen replied:
`maxTimeout=ZERO` seems to mean infinite, I wonder if the `TimeValue.min` here should respect that?

henningandersen reviewed (commented):
Looks good, left a couple comments.

ankikuma committed:
Add test

ankikuma committed:
Address review comments

ankikuma committed:
Fix gradle errors

ankikuma committed:
Add test

ankikuma committed:
Merge remote-tracking branch 'upstream/main' into feature/ES8850
Refresh to latest

ankikuma committed:
Add test

ankikuma committed:
Merge remote-tracking branch 'upstream/main' into feature/ES8850
Refresh to latest

ankikuma committed:
Cleanup Node.java

ankikuma commented:
I addressed the review comments and added a test. But I need some help with the test. The reindexing goes through inspite of the shutdown.

In fact, I tried shutting down all 3 nodes in my test - master only, coordinator only and data only. And the reindexing task still goes through. 

I even tried submitting the reindexing task after shutting down the 3 nodes and it still goes through!! I feel like I am missing some very basic understanding of how the integration tests work.

ankikuma committed:
Commit test

ankikuma committed:
Rewrite test

ankikuma committed:
Merge remote-tracking branch 'upstream/main' into feature/ES8850
Refresh to latest

ankikuma committed:
Modify test

ankikuma committed:
Fix test

ankikuma committed:
Adjust timeout defaults

ankikuma committed:
Cleanup

ankikuma committed:
Merge remote-tracking branch 'upstream/main' into feature/ES8850
Refresh to latest

ankikuma commented:
Added a test. Please take a look @henningandersen @Tim-Brooks whenever you get a chance.

I set the default timeout for the reindexing wait to be 10s. Please let me know if you think I should shorten it.

ankikuma committed:
Remove shutdown dependency from gradle

ankikuma committed:
Remove taskManager.java

ankikuma committed:
Refactor test

Tim-Brooks commented on server/src/main/java/org/elasticsearch/node/ShutdownFenceService.java:
```diff
@@ -0,0 +1,161 @@
+/**
+ * Extracted out the logic from {@link Node#prepareForClose()} to facilitate testing.
+ */
+public class ShutdownFenceService {
```
I know I was the one who created this name initially 🙃, but I'm not sure it is the best. Maybe lets change it to `ShutdownPreparationService`

ankikuma replied:
Sounds good :-)

Tim-Brooks reviewed (commented):
Mostly looks good to me. Just 1 comment.

ankikuma committed:
Merge remote-tracking branch 'upstream/main' into feature/ES8850
Refresh to latest

ankikuma committed:
Rename ShutdownFenceService to ShutdownCloseService

Tim-Brooks reviewed (approved):
LGTM

henningandersen commented on modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:
```diff
@@ -0,0 +1,149 @@
+
+        // Check for reindex task to appear in the tasks list and Immediately stop coordinating node
+        TaskInfo mainTask = findTask(ReindexAction.INSTANCE.name(), coordNodeName);
+        shutdownPrepareService.prepareForShutdown(taskManager);
```
I wonder why we need this, should the test not rely on the node stop doing this instead?

ankikuma replied:
Hi Henning, the stopNode() request does not go through the prepareForClose() codepath. I actually tried a "put shutdown" request first to try to exercise my code (without success) until Tim pointed out it doesn't go through prepareForClose(). That is the reason we had to refactor the logic from prepareForClose() into a separate service so we could test it from the IT.

There is this comment in prepareForClose():
"Note that this class is part of infrastructure to react to signals from the operating system - most graceful shutdown logic should use Node Shutdown, see {@link org.elasticsearch.cluster.metadata.NodesShutdownMetadata}."

Since prepareForClose() is triggered by OS signals, I think it is not called in the integration testing framework.

But please advice if you think that prepareForClose() will not be invoked from graceful shutdown codepaths. I do see that "forcefully" calling prepareForClose() is not the ideal way to test that we wait for reindexing tasks during shutdown.

henningandersen replied:
Thanks, this makes sense to me, sorry for the noise.

henningandersen commented on modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:
```diff
@@ -0,0 +1,149 @@
+    private static TaskInfo findTask(String actionName, String nodeName) {
+        ListTasksResponse tasks;
+        long start = System.nanoTime();
+        do {
```
Can we use assertBusy() instead?

ankikuma replied:
Sure.

henningandersen commented on server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java:
```diff
@@ -0,0 +1,167 @@
+
+            stopperRunner.accept("http-server-transport-stop", httpServerTransport::close);
+            stopperRunner.accept("async-search-stop", () -> awaitSearchTasksComplete(maxTimeout, taskManager));
+            stopperRunner.accept("reindex-stop", () -> awaitReindexTasksComplete(reindexTimeout, taskManager));
```
Is this the only change in this class? The refactor makes it a bit harder to reason about the change. We can keep it as is, but it might be beneficial to do the refactor separately for future PRs.

ankikuma replied:
The refactor was done for testing purpose as explained above.

henningandersen replied:
Yeah, I understand that. Still, is the line added (plus the method called) the only change here or has other changes sneaked in?

ankikuma replied:
Oh I see, I thought your question was about creating this new class. I only really added the awaitReindexTasksComplete() but in order to reuse the code being used by awaitSearchTasksComplete I created a new API waitForTasks() to replace [awaitSearchTasksComplete()](https://github.com/elastic/elasticsearch/blob/98d53352160d6d2c397ea74567ebc690b701973f/server/src/main/java/org/elasticsearch/node/Node.java#L664)

henningandersen reviewed (commented):
Looks good, left a few comments.

ankikuma commented on modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:
```diff
@@ -0,0 +1,149 @@
+        assertHitCount(prepareSearch(INDEX).setSize(0).setTrackTotalHits(true), numDocs);
+    }
+
+    private void checkDestinationIndex(String dataNodeName, int numDocs) {
```
Thanks. Done.

ankikuma commented on modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    private void createReindexTaskAndShutdown(final String coordNodeName) throws Exception {
+
```
Done.

henningandersen replied:
I still see it here?

ankikuma commented on modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:
```diff
@@ -0,0 +1,149 @@
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.debug("Encounterd " + e.toString());
```
Sure. I added an assertion error.

henningandersen replied:
Did you not push that yet?

We should probably use `fail(e)` instead (which is almost the same but shorter).

henningandersen reviewed (commented):
Left a few comments before realizing the changes were not pushed yet.

ankikuma committed:
Address Review comments

ankikuma commented:
I just pushed the changes. I wanted to wait to see if you would suggest a major revamp of the test :)

ankikuma committed:
code cleanup

ankikuma committed:
Merge remote-tracking branch 'upstream/main' into feature/ES8850
Refresh to latest

henningandersen commented on modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:
```diff
@@ -0,0 +1,145 @@
+        }, 10, TimeUnit.SECONDS));
+    }
+
+    private static void findTask(String actionName, String nodeName) throws Exception {
```
nit: I'd call this waitForTask instead.

henningandersen commented on modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:
```diff
@@ -0,0 +1,145 @@
+    private void checkDestinationIndex(String dataNodeName, int numDocs) {
+        assertTrue(indexExists(DEST_INDEX));
+        flushAndRefresh(DEST_INDEX);
+        assertTrue("Number of documents in source and dest indexes does not match", waitUntil(() -> {
+            final TotalHits totalHits = SearchResponseUtils.getTotalHits(
+                client(dataNodeName).prepareSearch(DEST_INDEX).setSize(0).setTrackTotalHits(true)
+            );
+            return totalHits.relation() == TotalHits.Relation.EQUAL_TO && totalHits.value() == numDocs;
+        }, 10, TimeUnit.SECONDS));
```
I think this can be something like following instead:

```
assertBusy(() -> assertHitCount(prepareSearch(DEST_INDEX).setSize(0).setTrackTotalHits(true), numDocs));
```

ankikuma replied:
Yes. Done.

henningandersen commented on modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:
```diff
@@ -0,0 +1,145 @@
+        assertBusy(() -> {
+            ListTasksResponse tasks = clusterAdmin().prepareListTasks(nodeName).setActions(actionName).setDetailed(true).get();
+            tasks.rethrowFailures("Find my task");
+            for (TaskInfo taskInfo : tasks.getTasks()) {
+                // Skip tasks with a parent because those are children of the task we want
+                assertFalse(taskInfo.parentTaskId().isSet());
+            }
```
I think this does not really wait. As a sidenote I am unsure if we need to wait and also whether skipping child tasks is important. But I think this should be:

```suggestion
            for (TaskInfo taskInfo : tasks.getTasks()) {
                // Skip tasks with a parent because those are children of the task we want
                if(taskInfo.parentTaskId().isSet()) == false) {
                    return;
                }
                fail("not found - yet");
            }
```

ankikuma replied:
You are right. Thanks for catching this. But I actually don't like using assertBusy() here because it waits for an increasing amount of time and even the first wait is 1 ms. So I do in fact want to change this back to what I had originally where we wait for 10 nanoseconds at a time.

ankikuma replied:
You are right. Thanks for catching this, I fixed it and tested it out as well (by holding off submission of the reindexing task).

henningandersen commented on modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexMetrics.java:
```diff
@@ -19,7 +19,7 @@ public class ReindexMetrics {
 
     public ReindexMetrics(MeterRegistry meterRegistry) {
-        this(meterRegistry.registerLongHistogram(REINDEX_TIME_HISTOGRAM, "Time to reindex by search", "millis"));
+        this(meterRegistry.registerLongHistogram(REINDEX_TIME_HISTOGRAM, "Time to reindex by search", "seconds"));
```
While I think I agree to the change, I'd prefer to do this in a separate PR.

henningandersen commented on server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java:
```diff
@@ -0,0 +1,171 @@
+public class ShutdownPrepareService {
+
+    private final Logger logger = LogManager.getLogger(ShutdownPrepareService.class);
+    Settings settings;
+    HttpServerTransport httpServerTransport;
+    TerminationHandler terminationHandler;
```
We should make all fields as final and private as possible. I think these can be private final?

ankikuma replied:
Yes. Thanks!

henningandersen commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -167,6 +156,12 @@ public class Node implements Closeable {
         Setting.Property.NodeScope
     );
 
+    public static final Setting<TimeValue> MAXIMUM_REINDEXING_TIMEOUT_SETTING = Setting.positiveTimeSetting(
```
I think we might as well move these two settings closer to their use, i.e., to `ShutdownPrepareService`.

ankikuma replied:
Sure done.

henningandersen commented on server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java:
```diff
@@ -0,0 +1,171 @@
+        }
+    }
+
+    private void waitForTasks(TimeValue timeout, String taskName, TaskManager taskManager) {
```
Looks like we have differing naming for similar things in this class, I'd prefer to name this: `awaitTasksComplete`, similar to the two other `await` methods.

ankikuma replied:
Yes, that's a better name. Thanks.

henningandersen reviewed (approved):
LGTM.

Left a number of smaller comments, but no need for another round.

ankikuma committed:
Review comments addressed

ankikuma commented:
Wait for reindexing tasks to finish during shutdown for an amount of time defined by settings. Also log the number of reindexing tasks still in flight after the wait.
